src/webgpu/util/texture/texelData.ts

```diff
@@ -26,263 +21,101 @@ export const enum TexelComponent {
-  'rg32float':              { ...repeatComponents(  kRG, float32), sRGB: false },
-  'rgba16uint':             { ...repeatComponents(kRGBA,  uint16), sRGB: false },
-  'rgba16sint':             { ...repeatComponents(kRGBA,  sint16), sRGB: false },
-  'rgba16float':            { ...repeatComponents(kRGBA, float16), sRGB: false },
-  'rgba32uint':             { ...repeatComponents(kRGBA,  uint32), sRGB: false },
-  'rgba32sint':             { ...repeatComponents(kRGBA,  sint32), sRGB: false },
-  'rgba32float':            { ...repeatComponents(kRGBA, float32), sRGB: false },
-
-  'rgb10a2unorm':           { componentOrder: kRGBA, componentInfo: { R: unorm10, G: unorm10, B: unorm10, A: unorm2 }, sRGB: false },
-  'rg11b10ufloat':          { componentOrder: kRGB, componentInfo: { R: float11, G: float11, B: float10 }, sRGB: false },
-  // TODO: the e5 is shared between all components; figure out how to write it.
-  'rgb9e5ufloat':           { componentOrder: kRGB, componentInfo: { R: componentUnimplemented, G: componentUnimplemented, B: componentUnimplemented }, sRGB: false },
-
-  'depth32float':           { componentOrder: [TexelComponent.Depth], componentInfo: { Depth: float32 }, sRGB: false },
-  'depth24plus':            { componentOrder: [TexelComponent.Depth], componentInfo: { Depth: null }, sRGB: false },
-  'depth24plus-stencil8':   { componentOrder: [TexelComponent.Depth, TexelComponent.Stencil], componentInfo: { Depth: null, Stencil: null }, sRGB: false },
-};
-
-export interface TexelDataRepresentation {
-  readonly componentOrder: TexelComponent[];
-  readonly componentInfo: TexelComponentInfo;
-
-  // Gets the data representation for |components| where |components| is the expected
-  // values when read in a shader. i.e. Passing in 1.0 for a 8-bit unorm component will
-  // yield 255.
-  getBytes(components: { [c in TexelComponent]?: number }): ArrayBuffer;
-
-  // Pack texel components into the packed byte representation. This may round values, but
-  // does not do unorm <-> float conversion.
-  packData(components: { [c in TexelComponent]?: number }): ArrayBuffer;
-
-  // Decode data into the shader representation
-  decode(components: { [c in TexelComponent]?: number }): { [c in TexelComponent]?: number };
-}
-
-class TexelDataRepresentationImpl implements TexelDataRepresentation {
-  constructor(
-    private readonly format: UncompressedTextureFormat,
-    readonly componentOrder: TexelComponent[],
-    readonly componentInfo: TexelComponentInfo,
-    private readonly sRGB: boolean
-  ) {}
-
-  private totalBitLength(): number {
-    return this.componentOrder.reduce((acc, curr) => {
-      return acc + this.componentInfo[curr]!.bitLength;
-    }, 0);
-  }
-
-  private writeTexelData(
-    data: ArrayBuffer,
-    bitOffset: number,
-    bitLength: number,
-    type: TexelWriteType,
-    value: number
-  ) {
+  components: PerTexelComponent<number>,
+  bitLengths: number | PerTexelComponent<number>,
```

Comment:
You could, but that would result in 4 overloads for a function like this? Or I guess it could be 2 to be simple, but the implementation code would be pretty much the same - like it still has to check the type of the arguments.

Reply:
You can't have real overloads; you can only provide multiple signatures for a single function. This is because JS has no overloading (of course, since there are no types) and TS won't do things like create mangled names to implement overloads.